### PR TITLE
Support PSR-17 StreamFactoryInterface

### PR DESCRIPTION
**Pull request by deguif**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no? 
| Deprecations?   | no
| Related tickets | https://github.com/php-http/HttplugBundle/pull/339
| Documentation   | 
| License         | MIT

#### What's in this PR?

Support for PSR-17 `StreamFactoryInterface`

### CODE REVIEW

**Comment by deguif** [comment]
Build is failing because PHP 5.4 is still supported (unsupported `::class` constant), maybe support for unmaintained PHP versions could be dropped?

**Comment by shulard** [comment]
I think you are right. The whole PHP-HTTP ecosystem is moving around the v2 so its time to deprecate some old PHP versions. Regarding that update, maybe supporting both HttplugFactory and PSR17Factory is not relevant anymore... PSR17 will be the de facto standard.

**Comment by dbu** [comment]
not just "de facto" but "officially accepted" standard :-)

afaik @Nyholm did a lot of thinking to have a clean upgrade path for the PSR, maybe he can chime in on this?

**Comment by deguif** [comment]
Yes, let's wait for his comment(s). 
I would be very interested to know about the upgrade path ;)

**Comment by dbu** [comment]
@Nyholm do you have input on this one?

**Comment by dbu** [comment]
ping @Nyholm

**Review by Nyholm — APPROVED** [review]
Im sorry for the awful delay. 

This looks super solid.

File: src/CachePlugin.php [hunk 1]
```diff
@@ -66,8 +67,12 @@ final class CachePlugin implements Plugin
      *              Defaults to an empty array
      * }
      */
-    public function __construct(CacheItemPoolInterface $pool, StreamFactory $streamFactory, array $config = [])
+    public function __construct(CacheItemPoolInterface $pool, $streamFactory, array $config = [])
     {
+        if (!($streamFactory instanceof StreamFactory) && !($streamFactory instanceof StreamFactoryInterface)) {
+            throw new \LogicException(\sprintf('StreamFactory must be an instance of %s or %s', StreamFactory::class, StreamFactoryInterface::class));
```
Comment (GrahamCampbell) [inline src/CachePlugin.php]: Why don't we throw a `TypeError` here?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/CachePlugin.php (hunk 1 above)
Comment (dbu) [reply src/CachePlugin.php]: oh good idea. @deguif could you change this to throw a TypeError instead?

Comment (deguif) [reply src/CachePlugin.php]: Yes I can change this ;)

Comment (deguif) [reply src/CachePlugin.php]: And ✅

File: src/CachePlugin.php [hunk 2]
```diff
@@ -69,8 +70,12 @@ final class CachePlugin implements Plugin
      *              Defaults to an empty array
      * }
      */
-    public function __construct(CacheItemPoolInterface $pool, StreamFactory $streamFactory, array $config = [])
+    public function __construct(CacheItemPoolInterface $pool, $streamFactory, array $config = [])
     {
+        if (!($streamFactory instanceof StreamFactory) && !($streamFactory instanceof StreamFactoryInterface)) {
+            throw new \TypeError(\sprintf('StreamFactory must be an instance of %s or %s.', StreamFactory::class, StreamFactoryInterface::class));
```
Comment (GrahamCampbell) [inline src/CachePlugin.php]: Should be:

```
throw new \TypeError(\sprintf('Argument 2 passed to %s::__construct() must be of type %s|%s, %s given.', self::class, StreamFactory::class, StreamFactoryInterface::class, gettype($streamFactory)));
```

To match the message given by PHP 8 for type errors with union types.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/CachePlugin.php (hunk 2 above)
Comment (deguif) [reply src/CachePlugin.php]: Even better: `\is_object($streamFactory) ? \get_class($streamFactory) : \gettype($streamFactory)` for the given type ;)

**Review by dbu — APPROVED** [review]
thanks a lot!

File: src/CachePlugin.php [hunk 3]
```diff
@@ -33,7 +34,7 @@ final class CachePlugin implements Plugin
     private $pool;
 
     /**
-     * @var StreamFactory
+     * @var StreamFactory|StreamFactoryInterface
```
Comment (dbu) [inline src/CachePlugin.php:37]: once https://github.com/php-http/message-factory/pull/38 is released, we can raise our version constraint to get that version and only use the psr interface from then on.

Comment (deguif) [reply src/CachePlugin.php:37]: Yes, and let ping me at this moment, I will be glad to do the PR ;)

File: src/CachePlugin.php (hunk 2 above)
Comment (GrahamCampbell) [reply src/CachePlugin.php]: Ah, yes. 👍

**Pull request merged**